File: Deniz/day_5.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        pre_arr = [1] * n
        post_arr = [1] * n
        res = [1] * n

        prefix = 1
        for i in range(n):
            pre_arr[i] = prefix
            prefix *= nums[i]

        postfix = 1
        for i in range(n-1, -1, -1):
            post_arr[i] = postfix
            postfix *= nums[i]

        for i in range(n):
            res[i] = pre_arr[i] * post_arr[i]

        return res

File: Ozan/day_5.java
// Solution for 238. Product of Array Except Self
// link: https://leetcode.com/problems/product-of-array-except-self/description/


class Solution {
    public int[] productExceptSelf(int[] nums) {
        int sizeOfArray = nums.length;
        int[] prefix = new int[sizeOfArray];
        int[] postfix = new int[sizeOfArray];

        int acc = 1;
        for (int i=0; i<sizeOfArray; i++) {
            acc *= nums[i];
            prefix[i] = acc;
        }

        acc = 1;
        for (int i=sizeOfArray-1; i>=0; i--) {
            acc *= nums[i];
            postfix[i] = acc;
        }

        int[] result = new int[sizeOfArray];
        for (int i=1; i<sizeOfArray-1; i++) {
            result[i] = prefix[i-1] * postfix[i+1];
        }

        result[0] = postfix[1];
        result[sizeOfArray-1] = prefix[sizeOfArray-2];

        return result;
    }
}